lib/ansible/module_utils/k8s_common.py

```diff
@@ -35,145 +44,703 @@
 except ImportError:
     HAS_YAML = False
 
+try:
+    import string_utils
+    HAS_STRING_UTILS = True
+except ImportError:
+    HAS_STRING_UTILS = False
+
+
+ARG_ATTRIBUTES_BLACKLIST = ('property_path',)
+PYTHON_KEYWORD_MAPPING = dict(zip(['_{0}'.format(item) for item in kwlist], kwlist))
+PYTHON_KEYWORD_MAPPING.update(dict([reversed(item) for item in iteritems(PYTHON_KEYWORD_MAPPING)]))
+
+ARG_SPEC = {
+    'state': {
+        'default': 'present',
+        'choices': ['present', 'absent'],
+    },
+    'force': {
+        'type': 'bool',
+        'default': False,
+    },
+    'resource_definition': {
+        'type': 'dict',
+        'aliases': ['definition', 'inline']
+    },
+    'src': {
+        'type': 'path',
+    },
+    'kind': {},
+    'name': {},
+    'namespace': {},
+    'description': {},
+    'display_name': {},
+    'api_version': {
+        'aliases': ['api', 'version']
+    },
+    'kubeconfig': {
+        'type': 'path',
+    },
+    'context': {},
+    'host': {},
+    'api_key': {
+        'no_log': True,
+    },
+    'username': {},
+    'password': {
+        'no_log': True,
+    },
+    'verify_ssl': {
+        'type': 'bool',
+    },
+    'ssl_ca_cert': {
+        'type': 'path',
+    },
+    'cert_file': {
+        'type': 'path',
+    },
+    'key_file': {
+        'type': 'path',
+    },
+}
+
+
+class AnsibleMixin(object):
+    _argspec_cache = None
+
+    @property
+    def argspec(self):
+        """
+        Introspect the model properties, and return an Ansible module arg_spec dict.
+        :return: dict
+        """
+        if self._argspec_cache:
+            return self._argspec_cache
+
+        argument_spec = copy.deepcopy(ARG_SPEC)
+        argument_spec.update(self.__transform_properties(self.properties))
+        self._argspec_cache = argument_spec
+        return self._argspec_cache
+
+    def object_from_params(self, module_params, obj=None):
+        """
+        Update a model object with Ansible module param values. Optionally pass an object
+        to update, otherwise a new object will be created.
+        :param module_params: dict of key:value pairs
+        :param obj: model object to update
+        :return: updated model object
+        """
+        if not obj:
+            obj = self.model()
+            obj.kind = string_utils.snake_case_to_camel(self.kind, upper_case_first=False)
+            obj.api_version = self.api_version.lower()
+        for param_name, param_value in iteritems(module_params):
+            spec = self.find_arg_spec(param_name)
+            if param_value is not None and spec.get('property_path'):
+                prop_path = copy.copy(spec['property_path'])
+                self.__set_obj_attribute(obj, prop_path, param_value, param_name)
+
+        if self.kind.lower() == 'project' and (module_params.get('display_name') or
```

Comment:
I think it's fine to keep this here for now but, I wonder if this check (and similar checks in the future) should go into the `OpenShift` client as they are specific to `OpenShift` and not `Kubernetes`.

Reply:
This is all Ansible module specific. What we're trying to do is move Ansible module code out of the OpenShift client, and into Ansible.

Obviously, though, a `project` is OpenShift specific, and we've given special handling to the project model, in order to make it easy for a user update the description and display_name attributes. 

It would be ideal, if we could move this logic completely into `openshift_common.py`. However, I don't see a way to subclass the AnsibleMixin and not essentially duplicate the whole `object_from_params` method. It just seems easier to have one `AnsibleMixin` for now. Maybe it's something we can iterate on in the future.

Reply:
@chouseknecht makes sense. Also, as a first iteration, this is already good so I wouldn't stress it to find a better abstraction at this stage. I think once this module matures, we can start thinking about splitting some of this logic.